### PR TITLE
Extend NvServices and implement IDirectory

### PR DESCRIPTION
**Pull request by bylaws**
IDirectory is used by many retail games to list the contents of directories, the NvServices functions added are all used by > 3.0.0 games.

### CODE REVIEW

File: app/src/main/cpp/skyline/vfs/directory.h [hunk 1]
```diff
@@ -19,17 +22,19 @@ namespace skyline::vfs {
         struct Entry {
             std::string name;
             EntryType type;
+            size_t size; //!< 0 if a directory
         };
 
         /**
          * @brief A descriptor for what will be returned when reading a directories contents
+         * @note MUST match HOS
```
Comment (PixelyIon) [inline app/src/main/cpp/skyline/vfs/directory.h:30]: This doesn't fit the rest of the codebase, a lot of structures need to match HOS and we don't explicitly mark them as such

File: app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_as_gpu.cpp [hunk 2]
```diff
@@ -43,32 +51,68 @@ namespace skyline::service::nvdrv::device {
...
 
+            if (data.flags.remap) {
+                auto region{regionMap.upper_bound(data.offset)};
+                if ((region == regionMap.begin()) || (region == regionMap.end())) {
+                    state.logger->Warn("Cannot remap an unmapped GPU address space region: 0x{:X}", data.offset);
+                    return NvStatus::BadParameter;
+                }
+
+                // Upper bound gives us the region after the one we want
+                region--;
```
Comment (PixelyIon) [inline app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_as_gpu.cpp:94]: Move the comment to the end of the line

File: app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_ctrl_gpu.cpp [hunk 3]
```diff
@@ -77,7 +77,13 @@ namespace skyline::service::nvdrv::device {
         if (data.gpuCharacteristicsBufSize < sizeof(GpuCharacteristics))
             return NvStatus::InvalidSize;
 
-        data.gpuCharacteristics = GpuCharacteristics{};
+        // The ioctl3 version of GetCharacteristics additionally outputs to the inline output buffer
```
Comment (PixelyIon) [inline app/src/main/cpp/skyline/services/nvdrv/devices/nvhost_ctrl_gpu.cpp:80]: `ioctl3` -> `IOCTL3`

File: app/src/main/cpp/skyline/vfs/directory.h [hunk 4]
```diff
@@ -19,17 +22,19 @@ namespace skyline::vfs {
...
          */
         union ListMode {
             struct {
                 bool directory : 1; //!< The directory listing will contain subdirectories
                 bool file : 1; //!< The directory listing will contain files
             };
-            u32 raw;
+            u32 raw{};
```
Comment (PixelyIon) [inline app/src/main/cpp/skyline/vfs/directory.h:37]: This doesn't affect anything unless you move it to above the `struct`

**Review by PixelyIon — CHANGES_REQUESTED** [review]

File: app/src/main/cpp/skyline/vfs/directory.h (hunk 1 above)
Comment (bylaws) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: Well this is part of vfs, other vfs structs don't need to match hos (see entry below). If you have a list of structs in other subsystems that need to match hos tell me and I'll add notes

File: app/src/main/cpp/skyline/vfs/directory.h (hunk 4 above)
Comment (bylaws) [reply app/src/main/cpp/skyline/vfs/directory.h:37]: nope, it does, it's just that unions only zero the first entry by default with {}. Adding {} makes it also zero out that as we discussed ages back. I did it this way as it allows you to use ListMode{} neatly.

File: app/src/main/cpp/skyline/vfs/directory.h (hunk 1 above)
Comment (PixelyIon) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: This applies to a lot of subsystems, it's readily apparent when structures (Esp. unions) are meant for the guest. This shouldn't just set aside that entire convention and need one here.

Comment (bylaws) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: Not really, anyone could assume it's not meant for the guest. Please tell me what subsystems you speak of so I can fix convention

File: app/src/main/cpp/skyline/vfs/directory.h (hunk 4 above)
Comment (PixelyIon) [reply app/src/main/cpp/skyline/vfs/directory.h:37]: You're wrong, the entire issue was that regardless of what you do it'll only init the first member in unions. You will need to change the order if you want the entire u32 to be 0-init'd.

Comment (bylaws) [reply app/src/main/cpp/skyline/vfs/directory.h:37]: You sure, I don't remember discussing that at all?

File: app/src/main/cpp/skyline/vfs/directory.h (hunk 1 above)
Comment (PixelyIon) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: The convention doesn't need fixing but essentially all services, kernel objects and GPU.

Comment (bylaws) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: all of them are obvious and all structs there are almost always guest, that isn't the case for vfs where most aren't for HOS

Comment (PixelyIon) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: Are you saying this isn't obvious with it being a union with a `raw` value and a bitfield struct, if you assume this is not for guest then I'm not sure what to say

Comment (bylaws) [reply app/src/main/cpp/skyline/vfs/directory.h:30]: it's definitely not obvious for EntryType, it's literally an enum

**Review by PixelyIon — APPROVED** [review]

**Pull request merged**